webdriver/tests/bidi/storage/get_cookies/partition.py

```diff
@@ -0,0 +1,230 @@
+    )
+
+    assert cookies == {
+        "cookies": [],
+        "partitionKey": {"sourceOrigin": source_origin_1},
+    }
+
+    await remove_cookie(bidi_session, new_tab["context"], cookie_name)
+
+
+@pytest.mark.parametrize("domain", ["", "alt"], ids=["same_origin", "cross_origin"])
+async def test_partition_context_iframe(
+    bidi_session, new_tab, inline, domain_value, origin, domain
+):
+    iframe_url = inline("<div id='in-iframe'>foo</div>", domain=domain)
+    page_url = inline(f"<iframe src='{iframe_url}'></iframe>")
+    await bidi_session.browsing_context.navigate(
+        context=new_tab["context"], url=page_url, wait="complete"
+    )
+
+    contexts = await bidi_session.browsing_context.get_tree(root=new_tab["context"])
+    iframe_context = contexts[0]["children"][0]
+
+    cookie_name = "foo"
+    cookie_value = "bar"
+    await add_cookie(bidi_session, iframe_context["context"], cookie_name, cookie_value)
+
+    # Check that added cookies are present on the right context
+    cookies = await bidi_session.storage.get_cookies(
+        partition=BrowsingContextPartitionDescriptor(iframe_context["context"])
+    )
+
+    assert cookies["partitionKey"] == {
+        "sourceOrigin": origin(protocol="https", domain=domain)
+    }
+    assert len(cookies["cookies"]) == 1
+    recursive_compare(
+        {
+            "domain": domain_value(domain=domain),
+            "httpOnly": False,
+            "name": cookie_name,
+            "path": "/webdriver/tests/support",
+            "sameSite": "none",
+            "secure": False,
+            "size": 6,
+            "value": {"type": "string", "value": cookie_value},
+        },
+        cookies["cookies"][0],
+    )
+
+    await remove_cookie(bidi_session, iframe_context["context"], cookie_name)
+
+
+@pytest.mark.parametrize(
+    "protocol",
+    [
+        "http",
+        "https",
+    ],
+)
+async def test_partition_source_origin(
+    bidi_session, top_context, new_tab, inline, domain_value, origin, protocol
+):
+    url_1 = inline("<div>foo</div>", domain="alt", protocol=protocol)
+    await bidi_session.browsing_context.navigate(
+        context=top_context["context"], url=url_1, wait="complete"
+    )
+    source_origin_1 = origin(protocol, "alt")
+
+    url_2 = inline("<div>bar</div>", protocol=protocol)
+    await bidi_session.browsing_context.navigate(
+        context=new_tab["context"], url=url_2, wait="complete"
+    )
+    source_origin_2 = origin(protocol)
+
+    cookie_name = "foo"
+    cookie_value = "bar"
+    await add_cookie(bidi_session, new_tab["context"], cookie_name, cookie_value)
+
+    # Check that added cookies are present on the right context
+    cookies = await bidi_session.storage.get_cookies(
+        partition=StorageKeyPartitionDescriptor(source_origin=url_2)
+    )
+
+    assert cookies["partitionKey"] == {"sourceOrigin": source_origin_2}
+    assert len(cookies["cookies"]) == 1
```

Comment:
I'm not sure why the cookie you set in line `await add_cookie(bidi_session, new_tab["context"], cookie_name, cookie_value)` should be in the partition container `sourceOrigin`. In case of Chromium implementation, if the set cookie is not partitioned, it is partition only be UserContext. Meaning all the non-partitioned cookies from and to all the domains falls under the same partition `{userContext: "..."}`. In our case the default `userContext` is emitted. 

When the cookie is requested from the partition `{sourceOrigin: "REQUIRED_SOURCE_ORIGIN"}`, only **partitioned** cookies set from `REQUIRED_SOURCE_ORIGIN` are retuned.

[**expand a storage partition spec**](https://w3c.github.io/webdriver-bidi/#expand-a-storage-partition-spec):
> 1. Let partition key be an empty [map](https://infra.spec.whatwg.org/#ordered-map).
> 4. For each name → default value in the [default values for storage partition key attributes](https://w3c.github.io/webdriver-bidi/#default-values-for-storage-partition-key-attributes):
>    1. Let value be partition spec[name] if it [exists](https://infra.spec.whatwg.org/#map-exists) or default value otherwise.
>    1. [Set](https://infra.spec.whatwg.org/#map-set) partition key[name] to value.
> 1. For each name in the remote end’s [required partition key attributes](https://w3c.github.io/webdriver-bidi/#required-partition-key-attributes):
>    1. If partition spec[name] [exists](https://infra.spec.whatwg.org/#map-exists):
>        1. [Set](https://infra.spec.whatwg.org/#map-set) partition key][name] to partition spec[name].
>    1. Otherwise:
>        1. Return [error](https://w3c.github.io/webdriver/#dfn-errors) with [error code](https://w3c.github.io/webdriver/#dfn-error-code) [underspecified storage partition](https://w3c.github.io/webdriver-bidi/#underspecified-storage-partition).
> 1. Return [success](https://w3c.github.io/webdriver/#dfn-success) with data partition key.

Reply:
@lutien

Reply:
Maybe the _partitioning_ in spec includes partitioning by `path`, `domain`, subdomains etc?

Reply:
Now as I have permissions to submit reviews, let's move the conversation in phabricator:  https://phabricator.services.mozilla.com/D199070